runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11797,6 +11797,338 @@ static TR::Register *inlineEncodeUTF16(TR::Node *node, TR::CodeGenerator *cg)
+      {
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, targetVector, 0x0f);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vsububm, node, permuteVector, targetVector, permuteVector);
+      }
+
+   // Set permuteVector = 0x00081018202830384048505860687078 (reversed for LE)
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, targetVector, 3);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vslb, node, permuteVector, permuteVector, targetVector);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vbpermq, node, searchVector, searchVector, permuteVector);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrwz, node, result, searchVector);
+
+   // Then count the number of leading zeroes from the obtained result. This tells us the index (in
+   // bytes) of the first matching element in the vector. Note that there is no way to count leading
+   // zeroes of a half-word, so we count leading zeroes of a word and subtract 16 since the value
+   // we're interested in is in the least significant half-word.
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::cntlzw, node, result, result);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, result, result, -16);
+
+   // Finally, combine this with the address of the last vector load to find the address of the
+   // first matching element in the string. Finally, use this to calculate the corresponding index.
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, result, result, currentAddress);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::subf, node, result, arrAddress, result);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, foundExactLabel);
+   if (!isLatin1)
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::srawi, node, result, result, 1);
+
+      {
+      TR::RegisterDependencyConditions *deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 14 + srm->numAvailableRegisters(), cg->trMemory());
+
+      deps->addPostCondition(arrAddress, TR::RealRegister::NoReg);
+      deps->addPostCondition(targetScalar, TR::RealRegister::NoReg);
+      deps->addPostCondition(startIndex, TR::RealRegister::NoReg);
+      deps->addPostCondition(endIndex, TR::RealRegister::NoReg);
+
+      deps->addPostCondition(cr0, TR::RealRegister::cr0);
+      deps->addPostCondition(cr6, TR::RealRegister::cr6);
+
+      deps->addPostCondition(zeroRegister, TR::RealRegister::NoReg);
+      deps->addPostCondition(result, TR::RealRegister::NoReg);
+      deps->addPostCondition(currentAddress, TR::RealRegister::NoReg);
+      deps->addPostCondition(endAddress, TR::RealRegister::NoReg);
+
+      deps->addPostCondition(targetVector, TR::RealRegister::NoReg);
+      deps->addPostCondition(targetVectorNot, TR::RealRegister::NoReg);
+      deps->addPostCondition(searchVector, TR::RealRegister::NoReg);
+      deps->addPostCondition(permuteVector, TR::RealRegister::NoReg);
+
```

Comment:
1) There are 4 registers in the dependency to be marked/set as excluding gr0, since they are sources of addi instructions. Something like:  deps->getPostConditions()->getRegisterDependency(index)->setExcludeGPR0();
     This will not cause any side effect, but guarantee the correctness, since you are using so many registers anyway (in case gr0 being free, it will be assigned to other registers not marked).

2) I am wondering if you can minimize the number of registers being used by checking if the children nodes' registers will not be reused.

3) I am looking at the code sequence to see whether I can spot something to account for the anomaly (7th-match vs. 8th-match).  I am suspecting short-latency VSX instructions (or/and  and  permute) being starved of the pipeline by long-latency instructions.  Not sure though.

Reply:
1. My bad, I meant to go through and mark registers as excluding gr0 as necessary before submitting but forgot. Will fix this.
2. I think it should be possible for me to check if `targetScalar`, `startIndex`, or `endIndex` won't be reused by checking the child refcount and donate them to the scratch register manager after use if that holds. I'm not sure whether I can do that for `arrObjectAddress` though, since it's marked as containing a collected reference and I don't know enough about how that works to say whether that would be safe or not.
3. As I've said before, the drop in speedup between 7th and 8th character matches has nothing to do with the new vectorized code becoming slower there and is instead caused by the old code unexpectedly becoming faster.

Reply:
2.  you can, since there is no GC point in that sequence (i.e  branch out to call another method, jitAsyncMessage or jitThrowException).  In addition, you are implicitly creating quite a few internal-pointers anyway (i.e. addresses pointing at the middle of object) ... that is un-safe, if they lived across GC points.

3. Ok, recall it now.  Make each match-point loop long enough ... and profile the two cases to see the tick distribution. That is the only way I can think of following it up.

Reply:
1 and 2 should be fixed now. As for 3, since the problem seems to be related to something about the scalar version, I think it's out of scope for this work and should be dealt with as an entirely separate matter (if it needs followup at all).

Reply:
One potential correctness issue:
  The same register(s) could be added into the dependency list multiple times.  The later code had better look like:
          if (....->referenceCount() > 1)
             {
             deps->addPostDependency() ...
             (setExcludeGPR0 if necessary)
             }

Another minor issue:
    I didn't see child(0) being evaluated.  I think child(0) is the String object itself(?).  If child(1) evaluation depends on the existence of child(0) result register, there might be a problem.  I am not sure though.

Reply:
for the case of referenceCount() <= 1,  the register(s) are in srm.  They will be added to deps with excluding GPR0 later anyway.

Reply:
Looked back at the code for `TR_PPCScratchRegisterManager` and it seems I got confused since `srm->stopUsingRegisters()` ignores donated registers but `srm->addScratchRegistersToDependencyList(...)` does not. I'll go ahead and fix that.

As for the 0th child, that child is simply a `JITHelpers` object on which the helper is being called. Its only reason for existing is to be checked against `null` to ensure that untrusted code cannot call these helpers unsafely since they can't get a reference to one.

Reply:
Register dependencies should be fixed now.

Reply:
Looks good ...